src/vdo.rs

```diff
@@ -0,0 +1,482 @@
+    pub sop_product_type_dfp: Pd3p2SopDfpProductType,
+    /// Connector Type
+    pub connector_type: Pd3p2IdHeaderVdoConnectorType,
+    /// USB Vendor ID
+    pub usb_vendor_id: u32,
+}
+
+impl FromBytes for Pd3p2IdHeaderVdo {
+    fn from_bytes(reader: &mut BitReader) -> Result<Self> {
+        let usb_host_capability = reader.read_bit()?;
+        let usb_device_capability = reader.read_bit()?;
+
+        let sop_product_type_ufp = reader.read(3)?;
+        let sop_product_type_ufp =
+            Pd3p2SopUfpProductType::n(sop_product_type_ufp).ok_or_else(|| Error::ParseError {
+                field: "sop_product_type_ufp".into(),
+                value: sop_product_type_ufp,
+                #[cfg(feature = "backtrace")]
+                backtrace: std::backtrace::Backtrace::capture(),
+            })?;
+
+        let modal_operation_supported = reader.read_bit()?;
+        let sop_product_type_dfp = reader.read(3)?;
+        let sop_product_type_dfp =
+            Pd3p2SopDfpProductType::n(sop_product_type_dfp).ok_or_else(|| Error::ParseError {
+                field: "sop_product_type_dfp".into(),
+                value: sop_product_type_dfp,
+                #[cfg(feature = "backtrace")]
+                backtrace: std::backtrace::Backtrace::capture(),
+            })?;
+
+        let connector_type = reader.read(2)?;
+        let connector_type =
+            Pd3p2IdHeaderVdoConnectorType::n(connector_type).ok_or_else(|| Error::ParseError {
+                field: "connector_type".into(),
+                value: connector_type,
+                #[cfg(feature = "backtrace")]
+                backtrace: std::backtrace::Backtrace::capture(),
+            })?;
+
+        reader.skip(5)?;
+
+        let usb_vendor_id = reader.read(16)?;
+        let hwdb = udev::Hwdb::new()?;
+        let modalias = format!("usb:v{:04X}*", usb_vendor_id);
+
+        let vendor_str = hwdb
+            .query(modalias)
+            .next()
+            .map_or(std::ffi::OsString::from("Unknown"), |entry| {
+                entry.name().to_os_string()
+            });
+
+        let vendor_str = CString::new(vendor_str.as_bytes())?;
+        let mut vendor = [0u8; 32];
+        let bytes = vendor_str.as_bytes_with_nul();
+        vendor[..bytes.len()].copy_from_slice(bytes);
+
+        Ok(Pd3p2IdHeaderVdo {
+            vendor,
+            usb_host_capability,
+            usb_device_capability,
+            sop_product_type_ufp,
+            modal_operation_supported,
+            sop_product_type_dfp,
+            connector_type,
+            usb_vendor_id,
+        })
+    }
+}
+
+#[repr(C)]
+#[derive(Debug, Clone, PartialEq, Default, Printf, Snprintf, N)]
+pub enum Pd3p2ProductTypeVdo {
+    /// See USBPDB 6.4.4.3.1.6
+    #[default]
+    PassiveCableVdo,
+    /// See USBPDB 6.4.4.3.1.7
+    ActiveCableVdo,
+    /// See USBPDB 6.4.4.3.1.9
+    VpdVdo,
+    /// See USBPDB 6.4.4.3.1.4
+    UfpVdo,
+    /// See USBPDB 6.4.4.3.1.5
+    DfpVdo,
+}
```

Comment:
The problem is that `cbindgen` simply does not understand namespaces, see https://github.com/mozilla/cbindgen/blob/master/docs.md#writing-your-c-api

```
🚨 NOTE: A major limitation of cbindgen is that it does not understand Rust's module system or namespacing. This means that if cbindgen sees that it needs the definition for MyType and there exists two things in your project with the type name MyType, it won't know what to do. Currently, cbindgen's behaviour is unspecified if this happens. However this may be ok if they have different cfgs
```

This is why all types in this crate are manually namespaced, including enum variants.

This *may*  be fixed with inline annotations to rename types, as in 

```
mod foo {
  /// cbindgen::rename="FooBar"
  struct Bar;
}
```

But I haven't tried it.

Reply:
I don't think there is a solution through cbindgen itself. What may work is to use newtypes and then expose those through the #[no_mangle] functions.

e.g.:

```
// We want this to translate to a typedef
#[repr(transparent)] 
struct Foo_Bar_Baz(pub foo::bar::Baz);

#[no_mangle]
pub extern "C" some_c_api(baz: &Foo_Bar_Baz);
```

We can probably automate this with a proc_macro that automatically builds the type's name by walking the module hierarchy. This can be gated behind the current "c_api" feature, thereby saving the Rust users from all this extra boilerplate.

@refi64 your opinion here?

Reply:
I think `cbindgen.toml`'s `export.rename` can do this?

Reply:
@refi64 I don't think it can rename types that have the same name. The syntax is "old_type" = "new_type", there is no way to write "foo::bar" = "FooBar", unless I've missed something.

Reply:
welp, apparently that's definitely a no:

https://github.com/mozilla/cbindgen/issues/7